File: submissions/team-PIXELZ/CoJourneyApp/src/utils/matching.ts
import { Journey, MatchFactor, MatchResult, User } from '../types';

/**
 * Clean and normalize text for simple location comparison
 */
function normalizeText(text: string): string {
  return (text || '').toLowerCase().trim().replace(/[^a-z0-9]/g, '');
}

/**
 * Parse time strings like "5:00 PM", "17:00", "5:30 pm", "8:00 AM" into minutes of day
 */
function parseTimeToMinutes(timeStr: string): number {
  if (!timeStr) return 0;
  const match = timeStr.match(/(\d+)(?::(\d+))?\s*(am|pm)?/i);
  if (!match) return 0;

  let hours = parseInt(match[1], 10);
  const minutes = match[2] ? parseInt(match[2], 10) : 0;
  const meridiem = match[3] ? match[3].toLowerCase() : null;

  if (meridiem === 'pm' && hours < 12) hours += 12;
  if (meridiem === 'am' && hours === 12) hours = 0;

  return hours * 60 + minutes;
}

/**
 * Rule-based Deterministic Matching Algorithm
 * Destination: 30%
 * Route overlap: 25%
 * Time: 20%
 * Preferences: 10%
 * Trust: 10%
 * History: 5%
 * Total: 100%
 */
export function calculateJourneyMatch(
  targetJourney: Partial<Journey>,
  candidateJourney: Journey,
  currentUser?: Partial<User>
): MatchResult {
  const reasons: string[] = [];
  const factors: MatchFactor[] = [];

  const targetDest = normalizeText(targetJourney.to || '');
  const candDest = normalizeText(candidateJourney.to || '');
  const targetFrom = normalizeText(targetJourney.from || '');
  const candFrom = normalizeText(candidateJourney.from || '');

  // 1. Destination Match (30 pts max)
  let destScore = 0;
  if (targetDest && candDest) {
    if (targetDest === candDest) {
      destScore = 30;
      reasons.push('Same destination');
    } else if (targetDest.includes(candDest) || candDest.includes(targetDest)) {
      destScore = 24;
      reasons.push('Nearby destination area');
    } else {
      destScore = 8;
    }
  } else {
    destScore = 20;
  }
  factors.push({
    label: 'Destination',
    score: destScore,
    maxScore: 30,
    description: destScore >= 24 ? 'High destination alignment' : 'Moderate destination distance',
  });

  // 2. Route Overlap (25 pts max)
  let routeScore = 0;
  const sameOrigin = targetFrom === candFrom && targetFrom !== '';
  if (sameOrigin && destScore >= 24) {
    routeScore = 25;
    reasons.push('Over 85% route overlap');
  } else if (sameOrigin) {
    routeScore = 20;
    reasons.push('Same departure point');
  } else if (
    targetFrom.includes(candFrom) ||
    candFrom.includes(targetFrom) ||
    targetDest.includes(candFrom) ||
    candDest.includes(targetFrom)
  ) {
    routeScore = 18;
    reasons.push('Along the same travel corridor');
  } else {
    routeScore = 14;
    reasons.push('Route overlaps along primary transit route');
  }
  factors.push({
    label: 'Route Overlap',
    score: routeScore,
    maxScore: 25,
    description: `${Math.round((routeScore / 25) * 100)}% route compatibility`,
  });

  // 3. Time Proximity (20 pts max)
  const targetMinutes = parseTimeToMinutes(targetJourney.time || '');
  const candMinutes = parseTimeToMinutes(candidateJourney.time || '');
  const diffMinutes = Math.abs(targetMinutes - candMinutes);

  let timeScore = 0;
  if (diffMinutes <= 15) {
    timeScore = 20;
    reasons.push(`Close departure time (±${diffMinutes} mins)`);
  } else if (diffMinutes <= 30) {
    timeScore = 16;
    reasons.push(`Similar travel window (±${diffMinutes} mins)`);
  } else if (diffMinutes <= 60) {
    timeScore = 12;
    reasons.push('Within 1 hour departure');
  } else {
    timeScore = 8;
  }
  factors.push({
    label: 'Travel Time',
    score: timeScore,
    maxScore: 20,
    description: `Difference: ${diffMinutes} minutes`,
  });

  // 4. Preferences Match (10 pts max)
  let prefScore = 10;
  if (candidateJourney.companionPreference && candidateJourney.companionPreference !== 'Any') {
    const userGender = currentUser?.gender || 'Male';
    if (candidateJourney.companionPreference === userGender) {
      prefScore = 10;
      reasons.push(`Matches ${candidateJourney.companionPreference} companion preference`);
    } else {
      prefScore = 4;
    }
  } else {
    prefScore = 10;
    reasons.push('Flexible cooperation preference');
  }
  factors.push({
    label: 'Preferences',
    score: prefScore,
    maxScore: 10,
    description: 'Compatible travel preferences',
  });

  // 5. Trust Score (10 pts max)
  const candTrust = candidateJourney.userTrustScore || 80;
  const trustScore = Math.round((Math.min(candTrust, 100) / 100) * 10);
  if (candidateJourney.userVerified) {
    reasons.push('Verified community member');
  }
  if (candTrust >= 90) {
    reasons.push('Highly trusted traveler');
  }
  factors.push({
    label: 'Trust & Reputation',
    score: trustScore,
    maxScore: 10,
    description: `${candTrust}/100 community trust`,
  });

  // 6. History / Past Cooperation (5 pts max)
  const historyScore = currentUser && currentUser.completedJourneys && currentUser.completedJourneys > 3 ? 5 : 4;
  if (historyScore === 5) {
    reasons.push('Strong cooperation track record');
  }
  factors.push({
    label: 'Cooperation History',
    score: historyScore,
    maxScore: 5,
    description: 'Active community participant',
  });

  const rawTotal = destScore + routeScore + timeScore + prefScore + trustScore + historyScore;
  const matchPercentage = Math.min(99, Math.max(45, rawTotal));

  return {
    journey: candidateJourney,
    matchPercentage,
    reasons,
    factors,
  };
}

/**
 * Score and sort candidate journeys by compatibility with target journey
 */
export function findJourneyMatches(
  targetJourney: Partial<Journey>,
  allJourneys: Journey[],
  currentUser?: Partial<User>
): MatchResult[] {
  return allJourneys
    .filter(j => j.id !== targetJourney.id && j.userId !== currentUser?.id)
    .map(candidate => calculateJourneyMatch(targetJourney, candidate, currentUser))
    .sort((a, b) => b.matchPercentage - a.matchPercentage);
}

File: backend/matching.py
import re

def normalize_text(text: str) -> str:
    return re.sub(r'[^a-z0-9]', '', (text or '').lower().strip())

def parse_time_to_minutes(time_str: str) -> int:
    if not time_str:
        return 0
    m = re.search(r'(\d+)(?::(\d+))?\s*(am|pm)?', time_str, re.IGNORECASE)
    if not m:
        return 0
    hours = int(m.group(1))
    minutes = int(m.group(2)) if m.group(2) else 0
    meridiem = m.group(3).lower() if m.group(3) else None

    if meridiem == 'pm' and hours < 12:
        hours += 12
    if meridiem == 'am' and hours == 12:
        hours = 0
    return hours * 60 + minutes

def calculate_cooperation_match(target: dict, candidate: dict, user: dict = None) -> dict:
    reasons = []

    t_dest = normalize_text(target.get("destination") or target.get("to") or "")
    c_dest = normalize_text(candidate.get("destination") or candidate.get("to") or "")
    t_from = normalize_text(target.get("start_location") or target.get("from") or "")
    c_from = normalize_text(candidate.get("start_location") or candidate.get("from") or "")

    # 1. Destination (30%)
    dest_score = 0
    if t_dest and c_dest:
        if t_dest == c_dest:
            dest_score = 30
            reasons.append("Same destination")
        elif t_dest in c_dest or c_dest in t_dest:
            dest_score = 24
            reasons.append("Nearby destination area")
        else:
            dest_score = 8
    else:
        dest_score = 20

    # 2. Route Overlap (25%)
    route_score = 0
    if t_from == c_from and t_from != "":
        route_score = 25 if dest_score >= 24 else 20
        reasons.append("Over 80% route overlap")
    elif t_from in c_from or c_from in t_from:
        route_score = 18
        reasons.append("Along the same travel corridor")
    else:
        route_score = 14
        reasons.append("Route overlaps along main highway")

    # 3. Time (20%)
    t_time = parse_time_to_minutes(target.get("travel_time") or target.get("time") or "")
    c_time = parse_time_to_minutes(candidate.get("travel_time") or candidate.get("time") or "")
    diff = abs(t_time - c_time)

    if diff <= 15:
        time_score = 20
        reasons.append(f"Similar travel time (±{diff} mins)")
    elif diff <= 30:
        time_score = 16
        reasons.append("Within 30 mins departure window")
    elif diff <= 60:
        time_score = 12
        reasons.append("Within 1 hour travel window")
    else:
        time_score = 8

    # 4. Preferences (10%)
    pref_score = 10
    reasons.append("Compatible cooperation preferences")

    # 5. Trust (10%)
    trust_val = candidate.get("trust_score") or candidate.get("user_trust_score") or 85
    trust_score = round((min(trust_val, 100) / 100.0) * 10)
    if candidate.get("verified", True):
        reasons.append("Verified community member")

    # 6. History (5%)
    history_score = 5
    reasons.append("Proven cooperation track record")

    total = min(99, max(45, dest_score + route_score + time_score + pref_score + trust_score + history_score))

    return {
        "match_percentage": total,
        "reasons": reasons,
        "candidate": candidate
    }
